### PR TITLE
[MANA-87] Implement function wrappers Nimrod requires

### PR DESCRIPTION
**Pull request by fyshhh**
This pull request implements a few functions that nimrod requires:
- `MPI_Comm_get_attr`
- `MPI_Comm_set_attr`
- `MPI_File_open`
- `MPI_File_close`
- `MPI_File_get_size`
- `MPI_File_read_at`
- `MPI_Waitany`
- `MPI_Get_elements_x`
- `MPI_Type_size_x`

And also implements incidental functions that are either closely related or will likely be used (for debugging):
- `MPI_Comm_delete_attr`
- `MPI_Error_class`
- `MPI_Error_string`
- `MPI_Get_elements`

It also implements an upper-half only method of storing attributes. We use the following structure:
`[ attribute key -> { [communicator -> attribute value], copy callback function, delete callback function, extra state }]`
We reimplement two corresponding functions `MPI_Comm_create_keyval` and `MPI_Comm_free_keyval`, and add callback function calls to `MPI_Comm_free`.

This pull request also adds test cases for noteworthy function wrappers added.

### CODE REVIEW

File: contrib/mpi-proxy-split/mpi-wrappers/mpi_unimplemented_wrappers.txt [hunk 1]
```diff
@@ -155,7 +149,6 @@ int MPI_Igather(const void *sendbuf, int sendcount, MPI_Datatype sendtype, void
 int MPI_Igatherv(const void *sendbuf, int sendcount, MPI_Datatype sendtype, void *recvbuf, const int recvcounts[], const int displs[], MPI_Datatype recvtype, int root, MPI_Comm comm, MPI_Request *request);
 int MPI_Get_address(const void *location, MPI_Aint *address);
 int MPI_Get_elements(const MPI_Status *status, MPI_Datatype datatype, int *count);
```
Comment (gc00) [inline contrib/mpi-proxy-split/mpi-wrappers/mpi_unimplemented_wrappers.txt:151]: Assuming that this PR  will add an implementation of `MPI_Get_elements` (see suggestion above), we should remove it here from `mpi_unimplementeed_wrappers.txt`.

File: contrib/mpi-proxy-split/mpi-wrappers/mpi_file_wrappers.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/****************************************************************************
+ *   Copyright (C) 2019-2021 by Gene Cooperman, Rohan Garg, Yao Xu          *
+ *   gene@ccs.neu.edu, rohgarg@ccs.neu.edu, xu.yao1@northeastern.edu        *
```
Comment (gc00) [inline contrib/mpi-proxy-split/mpi-wrappers/mpi_file_wrappers.cpp:3]: Illio,
    The convention is that each file should mention only the names of authors who made significant changes to the file.  Please don't hesitate to remove me, Rohan, and Yao, and replace all three with your own name.

In addition, for other files, if you just changed 2 or 3 lines, it's not important.  But where you made significant changes to a file (perhaps a large function, or 2 or more small functions), then please add your name to the copyright, in alphabetical order.

**Review by gc00 — APPROVED** [review]
Hi @fyshhh, I'm approving this in advance.  Please see my note on `MPI_Get_elements`.

Also, it doesn't have to be part of this PR.  But please later add some simple test code for communicator attributes.  This is the kind of code *data structures) where it's easy to have a bug.  This will give you added confidence that the bug is not here, when testing Nimrod.

And finally, please note my inline comment on adding your name to the copyright.

File: contrib/mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp [hunk 3]
```diff
@@ -168,26 +190,40 @@ USER_DEFINED_WRAPPER(int, Waitall, (int) count,
 }
 
 USER_DEFINED_WRAPPER(int, Waitany, (int) count,
-		     (MPI_Request *) array_of_requests, (int *) index,
-		     (MPI_Status *) status)
+                     (MPI_Request *) array_of_requests, (int *) index,
```
Comment (gc00) [inline contrib/mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:193]: We want to replace this TAB and others by 8 spaces.

Comment (gc00) [reply contrib/mpi-proxy-split/mpi-wrappers/mpi_request_wrappers.cpp:193]: @fyshhh, Has the TAB been replaced now?  If so, please mark this as "resolved".

**Comment by gc00** [comment]
@fyshhh, please do a `git pull --rebase origin master` and then resolve any conflicts in these commits.

For anyone else viewing this PR, we are waiting for a version of Nimrod with debug symbols to be made available, so that we can analyze one last bug.

**Comment by fyshhh** [comment]
As we move towards using `feature/dmtcp-master` as the `master` branch, this pull request has been deemed obsolete and its commits can be found in #114 instead.

**Pull request closed without merging**